File: frontend/lib/metrics.ts
import type { components } from "@/generated/api/schema";
import { formatDurationClock } from "@/lib/format";

type Schemas = components["schemas"];

export type SessionMetrics = Schemas["SessionMetricsRead"];
export type Metric = Schemas["MetricRead"];
export type MetricExplanation = Schemas["ExplanationRead"];
export type TimeInZone = Schemas["TimeInZoneRead"];
export type SessionLoad = Schemas["LoadRead"];
export type StrengthMetrics = Schemas["StrengthRead"];
export type DetectedInterval = Schemas["IntervalRead"];
export type AnchorPin = Schemas["AnchorPinRead"];
export type SessionStreams = Schemas["SessionStreamsRead"];
export type StreamChannel = Schemas["StreamChannel"];
export type LoadBasis = Schemas["LoadBasis"];
export type AnchorType = Schemas["AnchorType"];

/**
 * A metric slot, resolved into the one thing a component has to know.
 *
 * The API guarantees exactly one of `value` / `not_assessed` is set, and the
 * whole point of the shape is that a component branches **once**: it either
 * renders a number with its explanation or renders `NotAssessed` with the
 * reason, in the same slot either way (UI convention 4). This narrows the
 * union so that branch is a discriminated one rather than three null checks
 * that could disagree.
 */
export type ResolvedMetric =
  | {
      readonly kind: "value";
      readonly value: number;
      readonly explanation: MetricExplanation | null;
    }
  | { readonly kind: "absent"; readonly reason: string };

/** Narrow one metric slot. A malformed slot reads as absent, never as zero. */
export function resolve(metric: Metric | null | undefined): ResolvedMetric {
  if (metric && metric.value !== null && metric.value !== undefined) {
    return {
      kind: "value",
      value: metric.value,
      explanation: metric.explanation ?? null,
    };
  }
  return {
    kind: "absent",
    reason: metric?.not_assessed ?? "This metric was not computed.",
  };
}

/** The number, or `null`. For arithmetic; render through `resolve`. */
export function metricValue(metric: Metric | null | undefined): number | null {
  const resolved = resolve(metric);
  return resolved.kind === "value" ? resolved.value : null;
}

/** Round a metric to `digits`, or hand back the reason it has none. */
export function formatMetric(metric: Metric, digits = 0): string | null {
  const resolved = resolve(metric);
  return resolved.kind === "value" ? resolved.value.toFixed(digits) : null;
}

/**
 * A5.2's counterfactual, as a sentence — or `null` when there is no other
 * model to compare against.
 *
 * *"Load 79, from power. Had power been unavailable, the heart-rate model
 * would have given 75."* It is composed here rather than sent by the API
 * because it is a rendering of two numbers the API does send, and a server
 * that shipped the prose would be shipping copy the page cannot restyle.
 */
export function loadCounterfactual(load: SessionLoad): string | null {
  const power = load.power_load ?? null;
  const hr = load.hr_load ?? null;
  const basis = load.load_basis ?? null;
  if (power === null || hr === null || basis === null) {
    return null;
  }
  return basis === "power"
    ? `Had power been unavailable, the heart-rate model would have given ${Math.round(hr)}.`
    : `Had the heart-rate model been unavailable, power would have given ${Math.round(power)}.`;
}

/** How each load basis reads in a sentence. */
export const LOAD_BASIS_LABELS: Readonly<Record<LoadBasis, string>> = {
  power: "power",
  hr: "heart rate",
};

/**
 * The zone-ramp token one band of a zone model paints with.
 *
 * The seven-zone power model maps one-to-one onto the ramp. The five-zone
 * heart-rate model maps onto the **same** ramp so a zone bar reads the same
 * way whichever channel drew it — first and last zone take the endpoints and
 * the rest spread evenly, exactly as `globals.css` documents the rule
 * (Z1→zone-1, Z2→zone-2, Z3→zone-4, Z4→zone-5, Z5→zone-7).
 */
const HR_ZONE_RAMP: readonly number[] = [1, 2, 4, 5, 7];

export function zoneRampIndex(
  zoneModel: Schemas["ZoneModel"] | null,
  index: number,
): number {
  if (zoneModel === "lthr_5") {
    return HR_ZONE_RAMP[index - 1] ?? index;
  }
  return index;
}

/** One band of a rendered zone bar. */
export interface ZoneBand {
  readonly index: number;
  readonly name: string;
  readonly seconds: number;
  /** Share of the banded total, 0–1. Zero when nothing fell in the band. */
  readonly fraction: number;
  /** The `--color-zone-*` custom property this band is painted with. */
  readonly color: string;
  /** What the band says on hover: zone, time and share. */
  readonly title: string;
}

/**
 * A zone distribution, prepared for the bar.
 *
 * Every band of the model is kept, including the empty ones: a zone with no
 * time in it is a fact about the ride, and dropping it would make the bar's
 * shape depend on the data rather than on the model.
 */
export function zoneBands(distribution: TimeInZone): ZoneBand[] {
  const total = distribution.total_s ?? 0;
  return distribution.zones.map((zone) => {
    const fraction = total > 0 ? zone.seconds / total : 0;
    return {
      index: zone.index,
      name: zone.name,
      seconds: zone.seconds,
      fraction,
      color: `var(--color-zone-${zoneRampIndex(distribution.zone_model ?? null, zone.index)})`,
      title: `Z${zone.index} ${zone.name} · ${formatDurationClock(zone.seconds)} · ${(fraction * 100).toFixed(1)}%`,
    };
  });
}

/** How each anchor type is written in a pin line. */
export const PIN_UNITS: Readonly<Record<AnchorType, string>> = {
  ftp: "W",
  lthr: "bpm",
  max_hr: "bpm",
  resting_hr: "bpm",
  cp: "W",
  w_prime: "J",
};

/** The pin of one anchor type, or `undefined` when none was in force. */
export function pinOf(
  metrics: SessionMetrics,
  anchorType: AnchorType,
): AnchorPin | undefined {
  return metrics.pins.find((pin) => pin.anchor_type === anchorType);
}

/** One channel's column out of a stream payload, or `undefined`. */
export function channelValues(
  streams: SessionStreams,
  channel: StreamChannel,
): (number | null)[] | undefined {
  return streams.channels.find((entry) => entry.channel === channel)?.values;
}

/** What a drag-selection reports for the rows it covers. */
export interface SelectionStats {
  readonly durationS: number;
  readonly averagePower: number | null;
  readonly normalizedPower: number | null;
  readonly averageHr: number | null;
  readonly averageCadence: number | null;
}

/** The rolling window Coggan's normalized power is defined over, in seconds. */
export const NP_WINDOW_S = 30;

/**
 * Normalized power over a slice, by the same definition the backend uses.
 *
 * A deliberate second implementation, and the only one in this codebase:
 * a drag-selection is a range nobody asked the server about, and a round trip
 * per pixel of drag is not a design. It is the *same arithmetic* —
 * `app.domain.metrics.normalized_power`, 30 s trailing rolling mean, fourth
 * power, mean, fourth root, leading samples averaged over a shorter window —
 * and `lib/metrics.test.ts` pins it against the backend's own committed
 * fixtures so the two cannot drift silently. Nothing is ever stored from it:
 * the artefact's numbers are always the server's.
 */
export function normalizedPower(watts: readonly number[]): number | null {
  if (watts.length === 0) {
    return null;
  }
  let running = 0;
  let fourthPowerSum = 0;
  for (let index = 0; index < watts.length; index += 1) {
    running += watts[index];
    if (index >= NP_WINDOW_S) {
      running -= watts[index - NP_WINDOW_S];
    }
    const mean = running / Math.min(index + 1, NP_WINDOW_S);
    fourthPowerSum += mean ** 4;
  }
  return (fourthPowerSum / watts.length) ** 0.25;
}

function mean(values: readonly (number | null)[]): number | null {
  const present = values.filter((value): value is number => value !== null);
  return present.length === 0
    ? null
    : present.reduce((total, value) => total + value, 0) / present.length;
}

/**
 * Statistics for the rows a selection covers, computed in the browser.
 *
 * Null rows are **excluded**, never read as zero — the same rule the backend
 * applies, and the reason a selection across a coffee stop reports the riding
 * either side of it rather than an average dragged toward nothing.
 */
export function selectionStats(
  streams: SessionStreams,
  from: number,
  to: number,
): SelectionStats {
  const start = Math.max(0, Math.min(from, to));
  const end = Math.min(streams.length, Math.max(from, to) + 1);
  const slice = (channel: StreamChannel) =>
    (channelValues(streams, channel) ?? []).slice(start, end);
  const watts = slice("power").filter(
    (value): value is number => value !== null,
  );
  return {
    durationS: Math.max(0, end - start),
    averagePower: mean(slice("power")),
    normalizedPower: normalizedPower(watts),
    averageHr: mean(slice("hr")),
    averageCadence: mean(slice("cadence")),
  };
}

File: backend/app/api/schemas/metrics.py
"""Response schemas for a session's computed metrics and its stored streams.

Two resources, split by size. The **metrics** ride along on
``GET /sessions/{id}`` — a few dozen numbers with their explanations, which is
what the page header renders. The **streams** are their own endpoint because a
four-hour ride is 14 400 rows per channel and 1-2 MB on the wire, and no page
that merely lists sessions should pay for that.

**Every metric is one shape.** `MetricRead` carries a value with its
explanation *or* a `not_assessed` reason, never both and never neither. The UI
branches on that once — `NotAssessed` renders the reason in the slot the value
would have occupied — instead of inventing an empty state per number
(`.claude/rules/frontend-ui-conventions.md` rule 4).

The field names mirror `app.domain.session_analysis.analysis_to_json` exactly,
so the stored payload validates straight into these models. Extra keys are
ignored by default, which is what lets a later work package add a metric
without invalidating every artefact already written.
"""

import datetime as dt
import uuid

from pydantic import BaseModel, ConfigDict, Field

from app.api.validation import PostgresText
from app.domain.anchors import AnchorType, Provenance
from app.domain.metrics import LoadBasis
from app.domain.streams import AnomalyKind, StreamChannel
from app.domain.zones import ZoneModel
from app.persistence.metrics import MAX_REASON_LENGTH


class ExplanationRead(BaseModel):
    """Why a number is the number — data attached to the number, not page copy."""

    #: The arithmetic, written the way a human reads it.
    formula: str
    #: Named quantities that went in, already rendered. An anchor input names
    #: the **version's** value, provenance and effective date.
    inputs: dict[str, str]
    #: What the computation had to assume, in the order it assumed them.
    assumptions: list[str]
    #: Where the method comes from, when it comes from somewhere.
    citation: str | None


class MetricRead(BaseModel):
    """One metric slot: answered, or refused with a reason.

    Exactly one of ``value`` and ``not_assessed`` is non-null.
    ``explanation`` is present exactly when ``value`` is.
    """

    value: float | None = None
    explanation: ExplanationRead | None = None
    #: Why this metric has no value, in the athlete's terms and **naming the
    #: missing input** ("no heart rate was recorded"). Render it in the slot.
    not_assessed: str | None = None


def predates(metric: str) -> MetricRead:
    """The slot an artefact written before ``metric`` existed carries.

    A payload stored by an earlier version of the metric set simply has no key
    for a number added later, and the honest answer is neither a zero nor a
    silently missing field: it is "these numbers predate this one, recompute to
    get it" — which is a `not_assessed` reason like any other, rendered in the
    slot by the same component (`.claude/rules/frontend-ui-conventions.md` rule
    4) and fixed by the button already on the page.
    """
    return MetricRead(
        not_assessed=(
            f"these metrics were computed before {metric} was — recompute this "
            "session to add it"
        )
    )


class PowerMetricsRead(BaseModel):
    """Everything derived from the power channel."""

    normalized_power: MetricRead
    #: Work done **while moving**, over moving time (D194, D196) — the basis a
    #: head unit averages over, summed across the same seconds it divides by,
    #: and *not* the duration the load was computed over. Its explanation
    #: carries all of that; render it.
    average_power: MetricRead
    max_power: MetricRead
    intensity_factor: MetricRead
    #: NP over the mean of the **same recorded rows** (D196), never over the
    #: moving-time average power above: two divisors would let the ratio fall
    #: below 1, which no ride can do.
    variability_index: MetricRead
    work_kj: MetricRead
    work_above_ftp_kj: MetricRead
    #: Display only. Never subtracted from the duration load is computed over.
    coasting_time_s: MetricRead


class HeartRateMetricsRead(BaseModel):
    """Everything derived from the heart-rate channel."""

    average_hr: MetricRead
    max_hr: MetricRead
    hrss: MetricRead
    efficiency_factor: MetricRead


class CadenceMetricsRead(BaseModel):
    """Everything derived from the cadence channel."""

    average_cadence: MetricRead
    max_cadence: MetricRead


class SpeedMetricsRead(BaseModel):
    """Distance and speed, in km and km/h.

    Every field defaults to the "recompute me" reason so that an artefact
    written before these numbers existed still validates and still renders in
    its slot; see :func:`predates`.
    """

    #: Integrated from the speed channel, not from the GPS track.
    distance_km: MetricRead = Field(default_factory=lambda: predates("distance"))
    #: Over moving time, the same basis as average power (D194).
    average_speed_kmh: MetricRead = Field(
        default_factory=lambda: predates("average speed")
    )
    max_speed_kmh: MetricRead = Field(default_factory=lambda: predates("max speed"))


class TemperatureMetricsRead(BaseModel):
    """What the device's own sensor read, in degrees Celsius."""

    average_temp_c: MetricRead = Field(default_factory=lambda: predates("temperature"))
    min_temp_c: MetricRead = Field(default_factory=lambda: predates("temperature"))
    max_temp_c: MetricRead = Field(default_factory=lambda: predates("temperature"))


class ZoneTimeRead(BaseModel):
    """Seconds spent in one band of one zone model."""

    index: int
    name: str
    seconds: float


class TimeInZoneRead(BaseModel):
    """One channel's zone distribution, or the reason it has none."""

    #: Non-null when this channel produced no distribution; every field below
    #: is then null or empty.
    not_assessed: str | None = None
    #: The model the bands came from — pinned, because ``(anchor, model) ->
    #: zones`` is only deterministic while the model is recorded (A5.5).
    zone_model: ZoneModel | None = None
    #: One entry per band, ascending, **including the empty ones**: a zone
    #: with no time in it is a fact about the ride, and dropping it would make
    #: the bar's shape depend on the data.
    zones: list[ZoneTimeRead] = []
    #: Seconds that fell in any band — this channel's coverage, not the ride's
    #: elapsed time.
    total_s: float | None = None
    easy_s: float | None = None
    moderate_s: float | None = None
    hard_s: float | None = None
    #: Treff's PI over the three-zone split, or the reason it is degenerate.
    polarization_index: MetricRead | None = None
    explanation: ExplanationRead | None = None


class TimeInZoneBlockRead(BaseModel):
    """The zone distributions, one per channel that has one."""

    power: TimeInZoneRead
    hr: TimeInZoneRead


class LoadRead(BaseModel):
    """Both load models, the selected one, and the rule that chose it (A5.2).

    Both values are here whichever was selected — that is the whole point.
    The counterfactual sentence ("had power been unavailable, the HR model
    would have given 75") is composed by the client from these two fields.
    """

    not_assessed: str | None = None
    training_load: float | None = None
    load_basis: LoadBasis | None = None
    load_basis_rule: str | None = None
    power_load: float | None = None
    hr_load: float | None = None
    explanation: ExplanationRead | None = None


class StrengthRead(BaseModel):
    """What a strength session moved. Kilograms, **never** a training load."""

    not_assessed: str | None = None
    #: Σ ``reps × kg`` over the sets logged in kilograms. Never add this to
    #: ``training_load`` and never render the two in one column (v2 §5.4).
    volume_load_kg: float | None = None
    sets_completed: int | None = None
    #: Fraction of the logged sets that carried kilograms.
    coverage: float | None = None
    explanation: ExplanationRead | None = None


class IntervalRead(BaseModel):
    """One detected work interval, addressed by row on the 1 Hz grid."""

    start_index: int
    #: One past the last — ``[start, end)``.
    end_index: int
    duration_s: int
    average_power: float | None
    max_power: float | None
    average_hr: float | None


class AnchorPinRead(BaseModel):
    """An anchor version a metric artefact was computed against.

    Resolved rather than left as an id: the header renders `FTP 262 ± 15 ·
    estimated`, and a client that had to fetch four anchor versions to draw
    one line would fetch them on every session it showed.
    """

    anchor_type: AnchorType
    version_id: uuid.UUID
    value: float
    unit: str
    provenance: Provenance
    effective_date: dt.date
    ci_low: float | None
    ci_high: float | None


class SessionMetricsRead(BaseModel):
    """One version of one session's metrics, with what it was computed from."""

    #: 1-based position in the chain. A recompute writes n+1 and supersedes n;
    #: the old version stays readable.
    version: int
    computed_at: dt.datetime
    #: Why this version exists. Null on version 1.
    recompute_reason: str | None
    #: The anchor versions in force when this was computed (D115).
    pins: list[AnchorPinRead]
    power_zone_model: ZoneModel | None
    hr_zone_model: ZoneModel | None

    #: Elapsed minus every stop over 30 s — **the duration training load was
    #: computed over** (A4.4, A5.1).
    recording_time_s: float
    elapsed_time_s: float
    #: Rows of the cleaned speed column at or above 1 km/h, one second each
    #: (D196). The basis every *average* here is taken over (D194) — and the
    #: rows they are summed over — whenever the speed channel covered enough of
    #: the ride to be one; where it did not, this still reports what the column
    #: showed and each average's explanation names the recording time it fell
    #: back to. Never the load's duration term.
    moving_time_s: float
    #: Elapsed minus moving minus the seconds the speed channel had no reading
    #: for, derived server-side so a client never has to pick which durations
    #: to subtract — or mistake a sensor dropout for standing still.
    stopped_time_s: MetricRead = Field(default_factory=lambda: predates("stopped time"))

    power: PowerMetricsRead
    heart_rate: HeartRateMetricsRead
    cadence: CadenceMetricsRead
    speed: SpeedMetricsRead = Field(default_factory=SpeedMetricsRead)
    temperature: TemperatureMetricsRead = Field(default_factory=TemperatureMetricsRead)
    elevation_gain_m: MetricRead
    load: LoadRead
    time_in_zone: TimeInZoneBlockRead
    #: Deterministic from the stream alone, so it is versioned with the
    #: metrics rather than recomputed per consumer (D118).
    intervals: list[IntervalRead]
    strength: StrengthRead


# --- the stream payload -------------------------------------------------------


class StreamChannelRead(BaseModel):
    """One channel's cleaned column, on the 1 Hz grid."""

    channel: StreamChannel
    #: Which sensor produced it (A4.3), when the file said.
    source: str | None
    #: One entry per row, **nulls preserved**: a recording stop is a break in
    #: the trace, not a run of zeros, and a chart that filled them would draw
    #: a ride the athlete did not do.
    values: list[float | None]


class StreamAnomalyRead(BaseModel):
    """One region of one channel the cleaner repaired (A4.2)."""

    channel: StreamChannel
    start_index: int
    end_index: int
    kind: AnomalyKind
    substituted_value: float | None


class SessionStreamsRead(BaseModel):
    """The chart payload: every channel on one index-aligned grid.

    Separate from `SessionRead` because it is 1-2 MB for a long ride. Every
    column has exactly ``length`` entries by construction (A4.1), which is
    what lets a client index them together without checking.

    For a **merged** session (WP-6.5) this is the joined grid: the recordings
    laid end to end from the earliest one's origin, with the gap between them
    left as unrecorded rows and reported in ``recording_stops``. Every index
    here addresses that grid.
    """

    #: The first recording the samples came from, in time order — the only one
    #: for every session the MVP ingests.
    recording_id: uuid.UUID
    #: Every recording joined into this view, in time order. One entry unless
    #: the session was merged.
    recording_ids: list[uuid.UUID]
    #: The grid origin, aware UTC. Row ``i`` covers ``[t0 + i, t0 + i + 1)``.
    t0: dt.datetime
    #: Rows in the grid — the length of every channel's ``values``.
    length: int
    channels: list[StreamChannelRead]
    #: The pauses that were subtracted from recording time, as row ranges.
    recording_stops: list[StreamStopRead]
    #: The cleaner's repairs, so the chart can mark them. `resampled_only`
    #: certificates are **excluded**: nothing was repaired there.
    anomalies: list[StreamAnomalyRead]


class StreamStopRead(BaseModel):
    """One recording pause, as a half-open row range."""

    start_index: int
    end_index: int


class MetricsRecompute(BaseModel):
    """Why a recomputation was asked for. The body is optional.

    The reason lands on the **new** version (`recompute_reason`) and is what a
    later reader sees when two versions of one session's numbers disagree —
    "anchor changed" and "stream re-ingested" are different stories, and
    neither is reconstructible from the numbers themselves.
    """

    model_config = ConfigDict(extra="forbid")

    reason: PostgresText | None = Field(
        default=None,
        max_length=MAX_REASON_LENGTH,
        description="Why the metrics are being recomputed.",
    )
